[libpas] pas_local_view_cache_prepare_to_pop should ensure local_view_cache is committed

Constellation opened the pull request:
#### 5e744f6225b5145ef2ce4838b414d4258caf5431
<pre>
[libpas] pas_local_view_cache_prepare_to_pop should ensure local_view_cache is committed
<a href="https://bugs.webkit.org/show_bug.cgi?id=242539">https://bugs.webkit.org/show_bug.cgi?id=242539</a>

Reviewed by Mark Lam.

After 4c09a48ea introduced TLC decommit feature, pas_local_allocator and pas_local_view_cache can be decommitted
in the middle of the program. When we would like to get information from that, we need to keep in mind that we
need to ensure that the cache is committed first. While pas_local_allocator has the code committing it, pas_local_view_cache&apos;s
code for that is incomplete, leading to use of a broken pas_local_view_cache in an edge case.

When popping a pas_segregated_exclusive_view from pas_local_view_cache, we first check whether the cache has a view which can be popped
in pas_local_view_cache_prepare_to_pop. We had a fast path in this function that checks if `cache-&gt;top_index != cache-&gt;bottom_index` is true
to determine that this local_view_cache is active and not decommitted. But this is wrong: let&apos;s consider the following race conditions.

1. This local_view_cache is in the decommitted state. So at any time, the cache&apos;s backing page can be decommitted.
2. But top_index == bottom_index since it is already stopped. The key detail is that they may be non-zero even though they are the same.
3. We run `cache-&gt;top_index != cache-&gt;bottom_index`&apos;s `cache-&gt;top_index`.
4. Just after that, the scavenger runs and decommitted this backing page!
5. `cache-&gt;bottom_index` gets zero!
6. `cache-&gt;top_index != cache-&gt;bottom_index` evaluates to true even though the local_view_cache is actually cleared.

We should follow the protocol of commit for pas_local_allocator. This patch makes it aligned to pas_local_allocator&apos;s commit protocol so
that we can ensure that this local_view_cache is committed after this function.

* Source/bmalloc/libpas/src/libpas/pas_local_allocator_inlines.h:
(pas_local_allocator_refill_with_known_config):
* Source/bmalloc/libpas/src/libpas/pas_local_view_cache.h:
(pas_local_view_cache_prepare_to_pop):

Canonical link: <a href="https://commits.webkit.org/252312@main">https://commits.webkit.org/252312@main</a>
</pre>

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_local_allocator_inlines.h:
```diff
@@ -1074,17 +1074,27 @@ pas_local_allocator_refill_with_known_config(
+               changing active to stopped state, but does not prevent changing stopped state to decommitted state. In that case,
+               even just after setting is_in_use, it is possible that this becomes false since cache is already decommitted.
+               pas_local_view_cache_prepare_to_pop ensures that view_cache is committed. */
+            can_pop = pas_local_view_cache_prepare_to_pop(view_cache);
```
Do we need a pas_compiler_fence here after `prepare_to_pop`?

Constellation replied:
If the view is decommitted / stopped, we revive it in `pas_local_view_cache_prepare_to_pop`, and it requires taking a scavenger_lock. So `pas_local_allocator_scavenger_data_did_use_for_allocation` is ensured to be happen after `pas_local_view_cache_prepare_to_pop`.
If it is not revived, then, ordering does not matter since scavenger cannot make it stopped / decommitted.

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_local_view_cache.h:
```diff
@@ -94,8 +94,31 @@ static inline bool pas_local_view_cache_is_empty(pas_local_view_cache* cache)
+        return false;
+    }
+
+    /* Because kind is pas_local_allocator_view_cache_kind and is_in_use is already true,
+       this cache will not be decommitted. Safely use it */
+
```
Can you add a `PAS_TESTING_ASSERT(view_cache->scavenger_data.is_in_use)` at the top of this function to document this invariant?  This function should only be called while `is_in_use` is true.

I also suggest rephrasing this comment slightly:
```
/* Because kind is pas_local_allocator_view_cache_kind and is_in_use is already true,
   the scavenger will not decommit this cache. Hence, it is safe to use. */
```

Constellation replied:
In fact, it can be `false` (super fun part). Yeah, we set it to `true`. But it is possible that this local_view_cache's page is madvise-decommitted just after that. And OS could already collected this backing page. In that case, when accessing this field, it is zero page, so we will get `false`!
The invariant is that, after returning this function, `view_cache->scavenger_data.is_in_use` must be `true`. In `pas_local_allocator_scavenger_data_commit_if_necessary_slow`, we set it to `true` during reviving. I'll add this to the caller of this function.

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_local_view_cache.h:
```diff
@@ -94,8 +94,31 @@ static inline bool pas_local_view_cache_is_empty(pas_local_view_cache* cache)
+        /* It is stopped (it can be decommitted at any time now) or decommitted.
+           But we found that this local_view_cache will be used now. So, this commits the view to revive it for the next use.
+           And in either case (stopped or decommitted), it is already empty. So we cannot pop anything from this local_view_cache,
+           returning false. */
```
I suggest rephrasing this comment as:
```
/* The kind is either "stopped" or "decommitted" (which can allow the scavenger to decommit this cache at
   any time). This function is called because the caller wants to use this local_view_cache now. So, we need to
   commit the cache here to revive it for use. Regardless of whether the kind is "stopped" or "decommitted",
   the cache is already empty. So, return false to indicate that there's nothing to pop. */

Constellation replied:
Sounds good!

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_local_allocator_inlines.h:
```diff
@@ -1074,17 +1074,27 @@ pas_local_allocator_refill_with_known_config(
             pas_local_allocator_scavenger_data_did_use_for_allocation(&view_cache->scavenger_data);
 
-            if (pas_local_view_cache_prepare_to_pop(view_cache)) {
+            if (can_pop) {
```
Let's call this `has_view_to_pop` instead.

Constellation replied:
Changed.

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_local_view_cache.h:
```diff
@@ -94,8 +94,31 @@ static inline bool pas_local_view_cache_is_empty(pas_local_view_cache* cache)
+       to see that this is pas_local_allocator_view_cache_kind. Since decommit happens after setting it to pas_local_allocator_decommitted_kind,
+       we can ensure that (1) is_in_use is true, (2) this is not a stopped state, and (3) this will not be decommitted.
+       But if it is not, then this local_view_cache is stopped or decommitted and stopped state can be changed to decommitted state at any time.
+       In that case, we need to revive it to active state. */
```
I suggest rephrasing this comment as:
```
/* We would like to ensure that this pas_local_view_cache is committed before we return from this function.
   To ensure that, we first check if cache->scavenger_data.kind is pas_local_allocator_view_cache_kind i.e.
   the cache is in an "active" state. If so, then we know that the cache will remain committed because
   is_in_use is already true, and the state cannot be changed while it is "active" and is_in_use is true.

   Otherwise, if the pas_local_view_cache is not "active", the scavenger could have changed the state to
   "stopped" before is_in_use was set. Once "stopped", the scavenger may further change it to "decommitted"
   even if is_in_use is true. Once the cache is in the "decommitted" state, the scavenger may decommit the
   cache at any time. Hence, if the cache is not in the "active" state, we should assume it is possibly
   decommitted and need to revive it to the "active" state to ensure that the cache is committed before
   returning. */
```

Constellation replied:
Looks good!

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_local_allocator_inlines.h:
```diff
@@ -1074,17 +1074,27 @@ pas_local_allocator_refill_with_known_config(
+               it is possible that scavenger further runs and makes this cache even decommitted since is_in_use can prevent
+               changing active to stopped state, but does not prevent changing stopped state to decommitted state. In that case,
+               even just after setting is_in_use, it is possible that this becomes false since cache is already decommitted.
+               pas_local_view_cache_prepare_to_pop ensures that view_cache is committed. */
```
I suggest rephrasing this comment as:
```
/* Setting is_in_use = true here is insufficient by itself to prevent the cache from being decommitted.
   Consider this: just before we set is_in_use to true, the scavenger runs and puts the cache in the
   "stopped" state. Once the cache is "stopped", the is_in_use flag does not prevent the scavenger from
   further putting the view_cache in the "decommitted" state. Being in the "decommitted" state gives the
   scavenger license to decommit the cache. Calling pas_local_view_cache_prepare_to_pop here will ensure
   that the view_cache is committed. */
```

Constellation replied:
Sounds good! Changed.

MenloDorian reviewed (approved):
Nice fix. r=me.  I suggested some comment changes to hopefully organize the info a little better, and also to disambiguate all the "it"s.

Constellation reviewed (commented):
Thanks!

Constellation committed:
[libpas] pas_local_view_cache_prepare_to_pop should ensure local_view_cache is committed
https://bugs.webkit.org/show_bug.cgi?id=242539

Reviewed by Mark Lam.

After 4c09a48ea introduced TLC decommit feature, pas_local_allocator and pas_local_view_cache can be decommitted
in the middle of the program. When we would like to get information from that, we need to keep in mind that we
need to ensure that the cache is committed first. While pas_local_allocator has the code committing it, pas_local_view_cache's
code for that is incomplete, leading to use of a broken pas_local_view_cache in an edge case.

When popping a pas_segregated_exclusive_view from pas_local_view_cache, we first check whether the cache has a view which can be popped
in pas_local_view_cache_prepare_to_pop. We had a fast path in this function that checks if `cache->top_index != cache->bottom_index` is true
to determine that this local_view_cache is active and not decommitted. But this is wrong: let's consider the following race conditions.

1. This local_view_cache is in the decommitted state. So at any time, the cache's backing page can be decommitted.
2. But top_index == bottom_index since it is already stopped. The key detail is that they may be non-zero even though they are the same.
3. We run `cache->top_index != cache->bottom_index`'s `cache->top_index`.
4. Just after that, the scavenger runs and decommitted this backing page!
5. `cache->bottom_index` gets zero!
6. `cache->top_index != cache->bottom_index` evaluates to true even though the local_view_cache is actually cleared.

We should follow the protocol of commit for pas_local_allocator. This patch makes it aligned to pas_local_allocator's commit protocol so
that we can ensure that this local_view_cache is committed after this function.

* Source/bmalloc/libpas/src/libpas/pas_local_allocator_inlines.h:
(pas_local_allocator_refill_with_known_config):
* Source/bmalloc/libpas/src/libpas/pas_local_view_cache.h:
(pas_local_view_cache_prepare_to_pop):

Canonical link: https://commits.webkit.org/252312@main
